core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -212,6 +226,70 @@ class half_t {
   KOKKOS_FUNCTION
   half_t(unsigned long long rhs) : val(cast_to_half(rhs).val) {}
 
+  // Atomics
+  KOKKOS_FUNCTION
+  void atomic_add(const volatile half_t src) const volatile {
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    volatile float* lhs = (float*)&val;
+    volatile float* rhs = (float*)&src.val;
+    *lhs                = __half2float(val);
+    *rhs                = __half2float(src.val);
+    val                 = __float2half(lhs[0] + rhs[0]);
```

Comment:
> Is there a good reason to not just do

Yes. I want to perform volatile loads on the values at `&val` and `&src.val` not on the locals `lhs` and `rhs`

> Also, is casting to `volatile` actually sufficient for `atomic` operations? Wouldn't it be more appropriate to actually use CUDA's atomic interface (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicadd)?

No, we do not have truly atomic operations in in Kokkos_Cuda_Half.hpp. I thought about using the Cuda atomic interfaces with the `_system` postfix but they do not support `__half`. So, we may be able to use a 32-bit atomic add from Cuda's interfaces but we cannot provide a pointer to 4 bytes of space for this interface to write to; this is why I did the 32-bit arithmetic followed by a 16-bit atomic store.

Reply:
Yes, what I meant is to use a `float` `atomicAdd` on `lhs` and `rhs` instead of relying on `volatile` operations, basically I would have expected that the code would look more similar to the host code if `CUDA` doesn't provide the appropriate `atomics` directly.

Reply:
I see. We do have to ensure that the compiler sees the 32-bit loads & stores as volatile which is probably why it looks a bit different.